### ISSUE TITLE
Embedded platforms and the road to packaged components

### ISSUE THREAD

**Issue by balloob** [home-assistant/architecture#124]
So @rytilahti had a good point in his long (and mainly off-topic 😉) [comment](https://github.com/home-assistant/architecture/issues/121#issuecomment-448439297) to my [proposal to change how we deal with discovery](https://github.com/home-assistant/architecture/issues/121). But I felt that a) it should be it's own issue and b) it should go further. And credit where credit is due, it's also something @amelchio has said before, and probably others too.

@rytilahti proposed to make each platform a folder with their own metadata file. I've always shot this idea down, because it means that all our code has to be written twice, once for things based as a component and once for things based as a platform. That would be a maintenance hell and thus hinder progress.

However, thinking about this, I realized that we can embed the platforms in components. That way we forget about the idea of "platforms" altogether. Instead, an entity component can load embedded platforms from other components. 

File structure wise, we would go from:

```
hue/__init__.py
light/hue.py
```

to:

```
hue/__init__.py
hue/light.py
```

**The content of the file stays exactly the same**, but now we have all code for a single component in one folder.

This means a bunch of things, among others:

- Only one code path, it's now all components all the way down
- We can implement this backwards compatible, by just looking at both spots. Implementation wise, we just change [loader.py](https://github.com/home-assistant/home-assistant/blob/d42d8543c8d66a1ccd064bcc9b09092291ac00c3/homeassistant/loader.py#L54) to be defined as:
    ```python
    def get_platform(hass, domain, platform):
        return (
            get_component(hass, "{}.{}".format(platform, domain) or
            get_component(hass, "{}.{}".format(domain, platform))
    ```
- We can start having config entries for "platforms", because it's a component now.
- We can now start defining a metadata file for components per https://github.com/home-assistant/architecture/issues/121#issuecomment-448439297 (but discussing that is out of scope for this issue)
- Custom components will now be easier to distribute, which leads to…
- Packaged components. If we have a single folder for a component and all it's platforms, we can make sure we have a file with metadata to specify requirements (so we can install requirements before loading the Python file and thus import stuff at the top of our python files), versioning, code owners, logo, name. Now this can be independently distributed and updated. Not just by us, but by others too, like Hass.io add-on repositories including updates etc. Just like Hass.io add-ons, it will ease maintenance as no longer the core team is the bottleneck of all PR reviews (this is just a future vision, not to be discussed in this issue!)
- Entity components can be cleaned up and broken up into multiple files, like adding a `media_player/const.py` for the constants. This will help with circular dependencies like we're currently facing in https://github.com/home-assistant/home-assistant/pull/18700

Downsides:

The only one I can think of is that once all built-in platforms are embedded and we deprecate the old pattern of `light/hue.py`, we could break some custom components that didn't update yet.

**Comment by MartinHjelmare** [comment]
How are we thinking about services here? Will platform specific services be registered under the platform domain, eg light, or under the component domain, eg hue? Do we change anything about that with the new structure?

**Comment by balloob** [comment]
They should be published under the component domain. Hue already does this (`hue.hue_activate_scene`). Sonos does not (yet): `media_player.sonos_X`, but those should become `sonos.X`

**Comment by pvizeli** [comment]
I like that a lot because with config flow it was going to be weird. They need to be a component, and for a platform only piece before it was strange. This close this behaves.

**Comment by iantrich** [comment]
Can we elaborate on services? Are we saying that if I currently have the device `light.bed` but it is a x branded light, its service will be `x.x_turn_on_light` or am I missing something obvious?

**Comment by elupus** [comment]
To limit issues with circular dependencies the _init_.py must be mostly empty. Since importing component/x/const.py will automatically load _init_.py. thus we don't want that to pull in a bunch of stuff. 

PS. Im from the referenced wip branch above where I ran into problems with the constants for climate being defined inside climate platform, but needed from helpers.

**Comment by elupus** [comment]
@iantrich, I'm pretty sure that only applies to custom services. Not the default supplied by platform.

**Comment by balloob** [comment]
> Should we also consider separating entity components, and integration components?

I'm not ready for that yet. I want to keep it such that we have a single component interface on top of the core.

**Comment by elupus** [comment]
This will be lovely for custom components too btw. Since it allow:

cd custom_components
git clone component_a
git clone component_b

Ie multiple custom components hosted by somebody else. Now it's messy since they need to populate same directories.

**Comment by petri** [comment]
How about using [Python entry points](https://packaging.python.org/guides/creating-and-discovering-plugins/) as a plugin / component package discovery mechanism?

**Comment by rytilahti** [comment]
What are the remaining points that are to be discussed under this issue, assuming we are agreeing that the proposal sounds good enough to start tinkering with the code? I suppose the only (sofar) listed downside is not a major issue.

Should a new issue discussing the implementation be opened? The remark from @petri would be on-topic there (and IMO that is the way to go for the implementation, making (custom) components pypi installable is a nice side-effect of using an existing 'metadata' possibilities of setuptools).

**Comment by balloob** [comment]
You can open an issue and discuss MVP,  only represent functionality we have today in a manifest. Discovery is beyond MVP, as we first need to migrate all  of Home Assistant integrations to follow this issue and then a potential packaged component proposal, before we can even think about implementing it. This is the last comment about packaged components on this issue, all others will be removed.
